### PR TITLE
ecds: fix a flake in the integration test

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: gRPC mux is trying to do some work during destruction (drain the queue), so it needs the parent to stay alive.
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/filter/http/filter_config_discovery_impl.h [hunk 1]
```diff
@@ -125,6 +124,8 @@ class FilterConfigSubscription
   const std::string subscription_id_;
   absl::flat_hash_set<DynamicFilterConfigProviderImpl*> filter_config_providers_;
   friend class DynamicFilterConfigProviderImpl;
+
+  std::unique_ptr<Config::Subscription> subscription_;
```
Comment (mattklein123) [inline source/common/filter/http/filter_config_discovery_impl.h:128]: Thanks for the quick fix here. As this is a real bug and it's not immediately clear to me what the problem is, can you please do two things:
1) Add some comments here about ordering dependencies.
2) Can you add a unit test that would crash without this change?

Comment (kyessenov) [reply source/common/filter/http/filter_config_discovery_impl.h:128]: It's the stats scope. gRPC stream posts some stats on drainage, and it drains on destruction. Will do.

Comment (kyessenov) [reply source/common/filter/http/filter_config_discovery_impl.h:130]: It's another race. gRPC stream needs to stay alive on the server while the client closes it, to enable draining and the crash. Trying to find a way to emulate server closure while holding test alive.

File: test/integration/extension_discovery_integration_test.cc [hunk 2]
```diff
@@ -323,5 +327,17 @@ TEST_P(ExtensionDiscoveryIntegrationTest, BasicTwoSubscriptionsSameName) {
   EXPECT_EQ("200", response->headers().getStatusValue());
 }
 
+TEST_P(ExtensionDiscoveryIntegrationTest, DestroyDuringInit) {
```
Comment (mattklein123) [inline test/integration/extension_discovery_integration_test.cc:330]: Can you add more comments in here? It's not clear to me how rate limiting is part of this and what is going on here exactly. 

Also, like I said before, a unit test is fine if it's easier to test this case deterministically. Up to you.

Thanks a ton for working on this.

/wait

Comment (kyessenov) [reply test/integration/extension_discovery_integration_test.cc:330]: I have no idea why rate limiting is important, but it allows me to get through to the offending line here: https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_mux_impl.cc#L258. The block prior to that is guarded by rate limiting check.

Comment (mattklein123) [reply test/integration/extension_discovery_integration_test.cc:330]: OK this feels pretty fragile to me, but I would like to merge this to unflake things. If we can figure out a better way to test this vs. this as a follow up that would be good. cc @lizan who might have additional thoughts.

Comment (kyessenov) [reply test/integration/extension_discovery_integration_test.cc:330]: Yeah, I don't think stat update should be guarded by an exceptional event (rate limit kicking in). It creates a really tough edge case to hit.

**Pull request merged**